snapcraft/internal/repo/apt_sources_manager.py

```diff
@@ -0,0 +1,231 @@
+
+        try:
+            command = [
+                "sudo",
+                "install",
+                "--owner=root",
+                "--group=root",
+                "--mode=0644",
+                f_name,
+                str(dst_path),
+            ]
+            subprocess.run(command, check=True)
+        except subprocess.CalledProcessError:
+            raise RuntimeError(f"Failed to install repository config with: {command!r}")
+    finally:
+        os.unlink(f_name)
+
+
+def _get_host_arch() -> str:
+    return ProjectOptions().deb_arch
+
+
+class AptSourcesManager:
+    def __init__(
+        self, *, sources_list_d: pathlib.Path = pathlib.Path("/etc/apt/sources.list.d"),
+    ) -> None:
+        self._sources_list_d = sources_list_d
+
+    def _construct_deb822_source(
+        self,
+        *,
+        architectures: Optional[List[str]] = None,
+        components: Optional[List[str]] = None,
+        formats: Optional[List[str]] = None,
+        suites: List[str],
+        url: str,
+    ) -> str:
+        """Construct deb-822 formatted sources.list config string."""
+        with io.StringIO() as deb822:
+            if formats:
+                type_text = " ".join(formats)
+            else:
+                type_text = "deb"
+
+            print(f"Types: {type_text}", file=deb822)
+
+            print(f"URIs: {url}", file=deb822)
+
+            suites_text = " ".join(suites)
+            print(f"Suites: {suites_text}", file=deb822)
+
+            if components:
+                components_text = " ".join(components)
+                print(f"Components: {components_text}", file=deb822)
+
+            if architectures:
+                arch_text = " ".join(architectures)
+            else:
+                arch_text = _get_host_arch()
+
+            print(f"Architectures: {arch_text}", file=deb822)
+
+            return deb822.getvalue()
+
+    def _install_sources(
+        self,
+        *,
+        architectures: Optional[List[str]] = None,
+        components: Optional[List[str]] = None,
+        formats: Optional[List[str]] = None,
+        name: str,
+        suites: List[str],
+        url: str,
+    ) -> bool:
+        """Install sources list configuration.
+
+        Write config to:
+        /etc/apt/sources.list.d/snapcraft-<name>.sources
+
+        :returns: True if configuration was changed.
+        """
+        config = self._construct_deb822_source(
+            architectures=architectures,
+            components=components,
+            formats=formats,
+            suites=suites,
+            url=url,
+        )
+
+        if name not in ["default", "default-security"]:
+            name = "snapcraft-" + name
+
+        config_path = self._sources_list_d / f"{name}.sources"
+        if config_path.exists() and config_path.read_text() == config:
+            # Already installed and matches, nothing to do.
+            logger.debug(f"Ignoring unchanged sources: {config_path}")
```

Comment:
pylint recommends using lazy formatting in debug messages.

Reply:
indeed, fixed